### PR TITLE
[RISCV][GISel] Add calling convention support for half

### PR DESCRIPTION
**Pull request by dtcxzyw**

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+; RUN:    -global-isel -stop-after=irtranslator -verify-machineinstrs < %s \
+; RUN:   | FileCheck -check-prefix=RV32IZFH %s
+
+define half @callee_half_in_regs(half %x) nounwind {
+  ; RV32I-LABEL: name: callee_half_in_regs
+  ; RV32I: bb.1 (%ir-block.0):
+  ; RV32I-NEXT:   liveins: $x10
+  ; RV32I-NEXT: {{  $}}
+  ; RV32I-NEXT:   [[COPY:%[0-9]+]]:_(s32) = COPY $x10
+  ; RV32I-NEXT:   [[TRUNC:%[0-9]+]]:_(s16) = G_TRUNC [[COPY]](s32)
```
Comment (tschuett) [inline llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll:15]: I cannot read RISCV-V, but should this case be rejected? It looks as if it looses precision due to the trunk/anyext pair.

Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll:15]: > I cannot read RISC-V, but should this case be rejected? It looks as if it looses precision due to the trunk/anyext pair.

half-precision fp type only occupies 16-bits.

Comment (tschuett) [reply llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll:15]: Nevermind.

Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll:15]: I am wondering that whether we should handle nan-boxing here.
See https://dtcxzyw.github.io/riscv-isa-manual-host/unpriv-isa-asciidoc.html#nanboxing

File: llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll [hunk 2]
```diff
@@ -0,0 +1,136 @@
+; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=riscv32 -global-isel -stop-after=irtranslator \
+; RUN:    -verify-machineinstrs < %s \
+; RUN:   | FileCheck -check-prefix=RV32I %s
+; RUN: llc -mtriple=riscv32 -mattr=+zfh -target-abi ilp32f \
+; RUN:    -global-isel -stop-after=irtranslator -verify-machineinstrs < %s \
+; RUN:   | FileCheck -check-prefix=RV32IZFH %s
```
Comment (arsenm) [inline llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll:7]: Probably don't need -verify-machineinstrs

File: llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp [hunk 3]
```diff
@@ -340,7 +340,7 @@ static bool isSupportedArgumentType(Type *T, const RISCVSubtarget &Subtarget,
   // supported yet.
   if (T->isIntegerTy())
     return T->getIntegerBitWidth() <= Subtarget.getXLen() * 2;
-  if (T->isFloatTy() || T->isDoubleTy())
+  if (T->isHalfTy() || T->isFloatTy() || T->isDoubleTy())
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVCallLowering.cpp:343]: I still think this whole function should just be dropped and get the default treatment

File: llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  ; RV32IZFH-NEXT:   [[COPY4:%[0-9]+]]:_(s16) = COPY $f14_h
+  ; RV32IZFH-NEXT:   [[COPY5:%[0-9]+]]:_(s16) = COPY $f15_h
+  ; RV32IZFH-NEXT:   [[COPY6:%[0-9]+]]:_(s16) = COPY $f16_h
+  ; RV32IZFH-NEXT:   [[COPY7:%[0-9]+]]:_(s16) = COPY $f17_h
+  ; RV32IZFH-NEXT:   [[COPY8:%[0-9]+]]:_(s32) = COPY $x10
+  ; RV32IZFH-NEXT:   [[TRUNC:%[0-9]+]]:_(s16) = G_TRUNC [[COPY8]](s32)
+  ; RV32IZFH-NEXT:   $f10_h = COPY [[TRUNC]](s16)
+  ; RV32IZFH-NEXT:   PseudoRET implicit $f10_h
+  ret half %x
+}
```
Comment (arsenm) [inline llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll:136]: Didn't test the caller side

File: llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll [hunk 5]
```diff
@@ -0,0 +1,343 @@
+; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=riscv32 -global-isel -stop-after=irtranslator < %s \
+; RUN:   | FileCheck -check-prefix=RV32I %s
+; RUN: llc -mtriple=riscv32 -mattr=+zfh -global-isel -stop-after=irtranslator < %s \
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll:4]: Add another test with +f but not +zfh

File: llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+  ; RV32I-NEXT:   $x10 = COPY [[ANYEXT]](s32)
+  ; RV32I-NEXT:   PseudoRET implicit $x10
+  ;
+  ; RV32IF-LABEL: name: callee_half_in_regs
+  ; RV32IF: bb.1 (%ir-block.0):
+  ; RV32IF-NEXT:   liveins: $f10_f
+  ; RV32IF-NEXT: {{  $}}
+  ; RV32IF-NEXT:   [[COPY:%[0-9]+]]:_(s32) = COPY $f10_f
+  ; RV32IF-NEXT:   [[TRUNC:%[0-9]+]]:_(s16) = G_TRUNC [[COPY]](s32)
+  ; RV32IF-NEXT:   [[ANYEXT:%[0-9]+]]:_(s32) = G_ANYEXT [[TRUNC]](s16)
```
Comment (dtcxzyw) [inline llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll:26]: What can we do to remove these unnecessary trunc-anyext pairs?

Comment (topperc) [reply llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll:26]: They should be removed by the combiner, but we stopped before the combiner.

Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll:26]: Oh, ignore me :)

File: llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll (hunk 1 above)
Comment (topperc) [reply llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll:15]: SelectionDAG has this nan-boxing code in `RISCVTargetLowering::splitValueIntoRegisterParts`. I think its only used when f32 is legal.

```
  if (IsABIRegCopy && (ValueVT == MVT::f16 || ValueVT == MVT::bf16) &&           
      PartVT == MVT::f32) {                                                      
    // Cast the [b]f16 to i16, extend to i32, pad with ones to make a float      
    // nan, and cast to f32.                                                     
    Val = DAG.getNode(ISD::BITCAST, DL, MVT::i16, Val);                          
    Val = DAG.getNode(ISD::ANY_EXTEND, DL, MVT::i32, Val);                       
    Val = DAG.getNode(ISD::OR, DL, MVT::i32, Val,                                
                      DAG.getConstant(0xFFFF0000, DL, MVT::i32));                
    Val = DAG.getNode(ISD::BITCAST, DL, MVT::f32, Val);                          
    Parts[0] = Val;                                                              
    return true;                                                                 
  } 
```

Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll:15]: I think it should be handled during the legalization of G_FCONSTANT half. Can we merge this PR first?

Comment (dtcxzyw) [reply llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll:15]: Test case:
```
; llc -mtriple=riscv32 -mattr=+f -target-abi=ilp32 -verify-machineinstrs test.ll -o -
declare i32 @callee_half_in_regs(i32 %a, half %b) nounwind

define i32 @caller_half_in_regs() nounwind {
  %1 = call i32 @callee_half_in_regs(i32 1, half 2.0)
  ret i32 %1
}
```

Comment (topperc) [reply llvm/test/CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll:15]: > I think it should be handled during the legalization of G_FCONSTANT half. Can we merge this PR first?

I don't think that's what SelectionDAG does. It's handled during call handling. splitValueIntoRegisterParts isn't called just for constants.

**Review by topperc — APPROVED** [review]
LGTM

**Comment by dtcxzyw** [comment]
> Hi @dtcxzyw
> 
> If you compile with EXPENSIVE_CHECKS on and run the new testcase CodeGen/RISCV/GlobalISel/irtranslator/calling-conv-half.ll it fails with
> 
> ```
> *** Bad machine code: Copy Instruction is illegal with mismatching sizes ***
> - function:    caller_half_return_stack2
> - basic block: %bb.1  (0x55852fd10918)
> - instruction: $x10 = COPY %0:_(s16)
> Def Size = 32, Src Size = 16
> LLVM ERROR: Found 1 machine code errors.
> ```
> 
> Can also be seen if you just add `-verify-machineinstrs` to the last RUN line in the testcase.

Thank you! I will have a look.

**Pull request merged**